urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/common/number/IntervalType.java

```diff
@@ -0,0 +1,280 @@
+		return (IntervalNumber)zero;
+	}
+	public IntervalNumber one(){
+		return (IntervalNumber)one;
+	}
+
+	
+	private static IntervalType[] types = null;
+	private static int typeKey(IntervalType x){
+		if(types==null){
+			types = new IntervalType[]{IntegerInterval.type, LongInterval.type, DoubleInterval.type, APRInterval.type};
+		}
+		for(int i=0; i<types.length; i++){
+			if(x==types[i])
+				return i;
+		}
+		throw new IllegalArgumentException("unknown CohoType:  " + x.name());
+	}
+
+	public IntervalType promote(IntervalType that){
+		if(this==that)
+			return this;
+		int tk1 = typeKey(this), tk2 = typeKey(that);
+		int tk = Math.max(tk1,tk2);
+		return types[tk]; 
+	}
+	public static IntervalType promote(ScaleType t){
+		if(t==CohoBoolean.type || t==CohoInteger.type)
+			return IntegerInterval.type;
+		else if(t==CohoLong.type)
+			return LongInterval.type;
+		else if(t==CohoDouble.type)
+			return DoubleInterval.type;
+		else if(t==CohoAPR.type)
+			return APRInterval.type;
+		else
+			throw new IllegalArgumentException("Unknown subclass of CohoNumber: " + t);
+	}
+	
+	private static IntervalNumber promote(IntervalNumber a, IntervalType t){
+		if(t == a.type()){
+			return a;
+		}
+		if(t == IntegerInterval.type)
+			return IntegerInterval.create(a);
+		else if(t == LongInterval.type)
+			return LongInterval.create(a);
+		else if(t == DoubleInterval.type)
+			return DoubleInterval.create(a);
+		else if(t == APRInterval.type)
+			return APRInterval.create(a);
+		throw new IllegalArgumentException("Unknown subclass of CohoNumber: " + t);
+	}	
+	public static IntervalNumber promote(ScaleNumber a){
+		if(a.type()==CohoBoolean.type || a.type()==CohoInteger.type)
+			return IntegerInterval.create(a);
+		else if(a.type()==CohoLong.type)
+			return LongInterval.create(a);
+		else if(a.type()==CohoDouble.type)
+			return DoubleInterval.create(a);
+		else if(a.type()==CohoAPR.type)
+			return APRInterval.create(a);
+		else
+			throw new IllegalArgumentException("Unknown subclass of CohoNumber: " + a.type());
+	}
+	public static IntervalNumber promote(Number a){
+		if(a instanceof Integer || a instanceof Short || a instanceof Byte)
+			return IntegerInterval.create(a);
+		else if(a instanceof Long)
+			return LongInterval.create(a);
+		else if(a instanceof Double || a instanceof Float)
+			return DoubleInterval.create(a);
+		else 
+			return DoubleInterval.create(a);
+	}
+	
+
+	
+	public static IntervalNumber[] promote(IntervalNumber a, IntervalNumber b){
+		if(a.type()==b.type())
+			return new IntervalNumber[]{a,b};
+		IntervalType t = a.type().promote(b.type());
+		return new IntervalNumber[]{promote(a,t),promote(b,t)};
+	}
+	public static IntervalNumber[] promote(IntervalNumber a, ScaleNumber b){
+		return promote(a,promote(b));
+	}
+	public static IntervalNumber[] promote(IntervalNumber a, Number b){
+		return promote(a, promote(b));
+	}
+
+	public static IntervalNumber promoteOp(IntervalNumber n1, IntervalNumber n2, CohoNumber.ArithOp op){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 84 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.